server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/BlobManagerImpl.java

```diff
@@ -62,11 +71,29 @@ public BlobId toBlobId(MessageId messageId) {
 
     @Override
     public Blob retrieve(BlobId blobId, MailboxSession mailboxSession) throws MailboxException, BlobNotFoundException {
-        return getBlobFromAttachment(blobId, mailboxSession)
-                .orElseGet(() -> getBlobFromMessage(blobId, mailboxSession)
+        return getBlobFromUpload(blobId, mailboxSession)
+            .or(Throwing.supplier(() -> getBlobFromAttachment(blobId, mailboxSession)).sneakyThrow())
+            .orElseGet(() -> getBlobFromMessage(blobId, mailboxSession)
                 .orElseThrow(() -> new BlobNotFoundException(blobId)));
     }
 
+    private Optional<Blob> getBlobFromUpload(BlobId blobId, MailboxSession mailboxSession) {
+        if (blobId.getRawValue().startsWith(UPLOAD_PREFIX)) {
+            UploadId uploadId = UploadId.from(blobId.getRawValue().substring(UPLOAD_PREFIX.length()));
```

Comment:
`blobId.asUploadId()` ? It can be reuse in future

Reply:
https://github.com/apache/james-project/pull/584 did do just that :-P